File: ios/Sightread/Sightread/Services/GeminiVisionService.swift
import Foundation

struct GeminiVisionService: VisionAIService {
  let apiKey: String
  private let model = "gemini-2.0-flash"

  func analyze(jpegData: Data, prompt: String) async throws -> String {
    guard !apiKey.isEmpty else { throw VisionAIError.missingAPIKey }
    let base64 = jpegData.base64EncodedString()
    guard let url = URL(string: "https://generativelanguage.googleapis.com/v1beta/models/\(model):generateContent?key=\(apiKey)") else { throw VisionAIError.invalidResponse }
    let body: [String: Any] = ["contents": [["parts": [["text": prompt], ["inline_data": ["mime_type": "image/jpeg", "data": base64]]]]]]
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.httpBody = try JSONSerialization.data(withJSONObject: body)
    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
      let code = (response as? HTTPURLResponse)?.statusCode ?? -1
      throw VisionAIError.httpError(code, String(data: data, encoding: .utf8) ?? "")
    }
    guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          let candidates = json["candidates"] as? [[String: Any]],
          let content = candidates.first?["content"] as? [String: Any],
          let parts = content["parts"] as? [[String: Any]] else { throw VisionAIError.invalidResponse }
    let text = parts.compactMap { $0["text"] as? String }.joined()
    guard !text.isEmpty else { throw VisionAIError.invalidResponse }
    return text.trimmingCharacters(in: .whitespacesAndNewlines)
  }
}

File: android/Sightread/app/src/main/java/com/meta/wearable/dat/externalsampleapps/sightread/ai/GeminiVisionService.kt
package com.meta.wearable.dat.externalsampleapps.sightread.ai

import android.util.Base64
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

class GeminiVisionService(private val apiKey: String) : VisionAIService {
  override suspend fun analyze(jpegData: ByteArray, prompt: String): String =
      withContext(Dispatchers.IO) {
        require(apiKey.isNotBlank()) { "Add a Gemini API key in Settings." }
        val base64 = Base64.encodeToString(jpegData, Base64.NO_WRAP)
        val url =
            URL(
                "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=$apiKey"
            )
        val body =
            JSONObject()
                .put(
                    "contents",
                    JSONArray()
                        .put(
                            JSONObject()
                                .put(
                                    "parts",
                                    JSONArray()
                                        .put(JSONObject().put("text", prompt))
                                        .put(
                                            JSONObject()
                                                .put(
                                                    "inline_data",
                                                    JSONObject()
                                                        .put("mime_type", "image/jpeg")
                                                        .put("data", base64),
                                                ),
                                        ),
                                ),
                        ),
                )
        val conn = (url.openConnection() as HttpURLConnection).apply {
          requestMethod = "POST"
          setRequestProperty("Content-Type", "application/json")
          doOutput = true
          outputStream.use { it.write(body.toString().toByteArray()) }
        }
        val code = conn.responseCode
        val text =
            (if (code in 200..299) conn.inputStream else conn.errorStream)
                .bufferedReader()
                .readText()
        if (code !in 200..299) error("API error $code: $text")
        val json = JSONObject(text)
        val parts =
            json.getJSONArray("candidates")
                .getJSONObject(0)
                .getJSONObject("content")
                .getJSONArray("parts")
        buildString {
          for (i in 0 until parts.length()) {
            append(parts.getJSONObject(i).optString("text", ""))
          }
        }.trim()
      }
}
